lib/api/controllers/security.js

```diff
@@ -1480,6 +1333,176 @@ class SecurityController extends NativeController {
     });
     return securityDocument;
   }
+
+  /**
+   * @param {Request} request
+   * @param {{ _id: string, _kuzzle_info: object }} pojoUser
+   * @returns {Promise<unknown>}
+   * @private
+   */
+  async _persistUser(request, pojoUser) {
+    let loadedUser;
+
+    // Add metadata
+    pojoUser._kuzzle_info = {
+      author: request.context.user ? String(request.context.user._id) : null,
+      createdAt: Date.now(),
+      updatedAt: null,
+      updater: null
+    };
+
+    try {
+      loadedUser = await this.kuzzle.repositories.user.load(pojoUser._id);
+    }
+    catch (error) {
+      if (!(error instanceof NotFoundError)) {
+        throw error;
+      }
+      loadedUser = null;
+    }
+
+    if (loadedUser !== null) {
+      throw errorsManager.get('security', 'user', 'already_exists', pojoUser._id);
+    }
+
+    const strategies = request.input.body.credentials
+      ? Object.keys(request.input.body.credentials)
+      : [];
+
+    // Early checks before the user is created
+    for (const strategy of strategies) {
+      if (!this.kuzzle.pluginsManager.listStrategies().includes(strategy)) {
+        throw errorsManager.get('security', 'credentials', 'unknown_strategy', strategy);
+      }
+
+      if (
+        await this.kuzzle.pluginsManager.getStrategyMethod(strategy, 'exists')(
+          request,
+          pojoUser._id,
+          strategy
+        )
+      ) {
+        throw errorsManager.get(
+          'security',
+          'credentials',
+          'database_inconsistency',
+          pojoUser._id);
+      }
+    }
+
+    const options = {
+      database: {
+        method: 'create',
+        refresh: getRefresh(request)
+      }
+    };
+
+    // Throw in case of failure with the right KuzzleError object
+    const createdUser = await this.kuzzle.repositories.user
+      .fromDTO(pojoUser)
+      .then(modifiedUser =>
+        this.kuzzle.repositories.user.persist(modifiedUser, options)
+      )
+      .then(modifiedUser => formatProcessing.serializeUser(modifiedUser));
+
+    // Creating credentials
+    let creationFailure = null;
+
+    for (const strategy of strategies) {
```

Comment:
Nitpicking: this could be replaced by
```suggestion
    for (const [strategyName, StrategieContent] of Object.entries(strategies)) {
```

Reply:
Same here: `strategies` is the result of an `Object.keys` so I don't think your suggestion does what you want it to do.

Reply:
My point was to replace `strategies` by the content of `request.input.body.credentials` or an empty object.
But I'm not gonna block this PR for that :+1: